runtime.md

```diff
@@ -67,38 +62,55 @@ Note: these operations are not specifying any command-line APIs, and the paramen
+In some implementations this means that a temporary process is created in the PID namespace but it pauses until the `start` operation is invoked before replacing the process with the user-specified code.
+
+Any error generated during the creation of the container MUST generate an error and a new container MUST not be created.
+
+Runtime callers who are interested in pre-create validation can run [bundle-validation tools](implementations.md#testing--tools) before invoking the start operation.
+
+### Start
+`start <container-id>`
 
-The runtime MAY validate `config.json` against this spec, either generically or with respect to the local system capabilities, before creating the container ([step 2](#lifecycle)).
-If the runtime does not perform initial validation and triggers an error due to an invalid or incompatible configuration, it MUST generate an error and jump to cleanup ([step 7](#lifecycle)).
-Runtime callers who are interested in pre-start validation can run [bundle-validation tools](implementations.md#testing--tools) before invoking the start operation.
+This operation MUST generate an error if it is not provided the container ID.
+Attempting to start a container that does not exist MUST generate an error.
+Attempting to start an already started container MUST have no effect on the container and MUST generate an error.
+This operation MUST run the user-specified code as specified by [`process`](config.md#process-configuration).
+If the runtime fails to run code as specified, an error MUST be generated.
+This process MUST be run in the PID namespace associated with the container.
 
-Attempting to start an already running container MUST have no effect on the container and MUST generate an error.
+### Kill
+`kill <container-id> <signal>`
 
-### Stop
+This operation MUST generate an error if it is not provided the container ID.
+Attempting to send a signal to a container that is not running MUST have no effect on the container and MUST generate an error.
+This operation MUST send the specified signal to the process in the container.
 
-`stop <container-id>`
+### Delete
+`delete <container-id>`
 
 This operation MUST generate an error if it is not provided the container ID.
-This operation MUST stop and delete a running container.
-Stopping a container MUST stop all of the processes running within the scope of the container.
-Deleting a container MUST delete the associated namespaces and resources associated with the container.
-Once a container is deleted, its `id` MAY be used by subsequent containers.
-Attempting to stop a container that is not running MUST have no effect on the container and MUST generate an error.
+Attempting to delete a container that does not exist MUST generate an error.
+Attempting to delete a container whose process is still running MUST first stop the process by invoking the `kill` operation on the container.
```

Comment:
Since kill is asynchronous (on Linux, at least) and the delete command will be neither the parent of the container process nor its subreaper (so can't wait on it) is it possible for the delete to efficiently do an implicit stop? Is having kill implicit worth needing the implementation of delete to loop waiting for the container to exit? What happens if the container process is zombied, can/should delete time out in this case?

If there isn't a simple solution I've missed, maybe just having delete error if the container hasn't exited (and having a separate "non-optional kill" command) is cleaner? (I think that also matches current runC behaviour)